src/dhcp/aca_dhcp_state_handler.cpp

```diff
@@ -0,0 +1,139 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aca_log.h"
+#include "aca_util.h"
+#include "aca_dhcp_server.h"
+#include "aca_dhcp_state_handler.h"
+#include "aca_goal_state_handler.h"
+#include "goalstateprovisioner.grpc.pb.h"
+#include <future>
+
+using namespace aca_dhcp_programming_if;
+using namespace alcor::schema;
+
+namespace aca_dhcp_state_handler
+{
+Aca_Dhcp_State_Handler::Aca_Dhcp_State_Handler()
+{
+  ACA_LOG_INFO("DHCP State Handler: using dhcp server\n");
+  this->dhcp_programming_if = &(aca_dhcp_server::ACA_Dhcp_Server::get_instance());
+
+  int rc = this->dhcp_programming_if->initialize();
+
+  if (rc == EXIT_SUCCESS) {
+    ACA_LOG_INFO("DHCP Programming initialization succeed\n");
+  } else {
+    ACA_LOG_ERROR("DHCP Programming initialization failed\n");
+    throw std::system_error(ENXIO, std::generic_category(),
+                            "DHCP Programming initialization failed\n");
+  }
+}
+
+Aca_Dhcp_State_Handler::~Aca_Dhcp_State_Handler()
+{
+  // allocated dhcp_programming_if is destroyed when program exits.
+}
+
+Aca_Dhcp_State_Handler &Aca_Dhcp_State_Handler::get_instance()
+{
+  // It is instantiated on first use.
+  // allocated instance is destroyed when program exits.
+  static Aca_Dhcp_State_Handler instance;
+  return instance;
+}
+
+int Aca_Dhcp_State_Handler::update_dhcp_state_workitem(const DHCPState current_DhcpState,
+                                                       GoalStateOperationReply &gsOperationReply)
+{
+  dhcp_config stDhcpCfg;
+  int overall_rc = EXIT_SUCCESS;
+  ulong culminative_dataplane_programming_time = 0;
+  ulong culminative_network_configuration_time = 0;
+
+  auto operation_start = chrono::steady_clock::now();
+
+  DHCPConfiguration current_DhcpConfiguration = current_DhcpState.configuration();
+  stDhcpCfg.mac_address = current_DhcpConfiguration.mac_address();
+  stDhcpCfg.ipv4_address = current_DhcpConfiguration.ipv4_address();
+  stDhcpCfg.ipv6_address = current_DhcpConfiguration.ipv6_address();
+  stDhcpCfg.port_host_name = current_DhcpConfiguration.port_host_name();
+
+  switch (current_DhcpState.operation_type()) {
+  case OperationType::CREATE:
+    overall_rc = this->dhcp_programming_if->add_dhcp_entry(&stDhcpCfg);
+    break;
+  case OperationType::UPDATE:
+    overall_rc = this->dhcp_programming_if->update_dhcp_entry(&stDhcpCfg);
+    break;
+  case OperationType::DELETE:
+    overall_rc = this->dhcp_programming_if->delete_dhcp_entry(&stDhcpCfg);
+    break;
+  default:
+    ACA_LOG_DEBUG("=====>wrong dhcp operation\n");
+    return EXIT_FAILURE;
```

Comment:
We want to continue here after setting the overall_rc to EXIT_FAILURE instead of returning here. The reason is we need to complete the remaining code below to add_goal_state_operation_status to reply the failing status back to data plane manager.